src/Compilers/Core/MSBuildTask/MapSourceRoots.cs

```diff
@@ -0,0 +1,229 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Tasks;
+using Microsoft.Build.Utilities;
+
+namespace Microsoft.CodeAnalysis.BuildTasks
+{
+    /// <summary>
+    /// Given a list of SourceRoot items produces a list of the same items with added <c>MappedPath</c> metadata that
+    /// contains calculated deterministic source path for each SourceRoot.
```

Comment:
Consider adding some details:
"... with added MappedPath metadata. The MappedPath is either
- the path (ItemSpec) itself, when non-deterministic,
- a calculated deterministic source path ('/_/', '/_1', etc) for each SourceRoot, when deterministic."

Reply:
Added